mypy/build.py

```diff
@@ -778,95 +785,188 @@ def is_file(path: str) -> bool:
-        # and x.py* exist.  How?  We sort the directory items, so x
-        # comes before x.py and x.pyi.  But the preference for .pyi
-        # over .py is encoded in find_module(); even though we see
-        # x.py before x.pyi, find_module() will find x.pyi first.  We
-        # use hits to avoid adding it a second time when we see x.pyi.
-        # This also avoids both x.py and x.pyi when x/ was seen first.
-        hits = set()  # type: Set[str]
-        for item in sorted(os.listdir(os.path.dirname(module_path))):
-            abs_path = os.path.join(os.path.dirname(module_path), item)
-            if os.path.isdir(abs_path) and \
-                    (os.path.isfile(os.path.join(abs_path, '__init__.py')) or
-                    os.path.isfile(os.path.join(abs_path, '__init__.pyi'))):
-                hits.add(item)
-                result += find_modules_recursive(module + '.' + item, lib_path)
-            elif item != '__init__.py' and item != '__init__.pyi' and \
-                    item.endswith(('.py', '.pyi')):
-                mod = item.split('.')[0]
-                if mod not in hits:
-                    hits.add(mod)
-                    result += find_modules_recursive(
-                        module + '.' + mod, lib_path)
-    return result
-
-
-def verify_module(id: str, path: str) -> bool:
-    """Check that all packages containing id have a __init__ file."""
-    if path.endswith(('__init__.py', '__init__.pyi')):
-        path = dirname(path)
-    for i in range(id.count('.')):
-        path = dirname(path)
-        if not any(os.path.isfile(os.path.join(path, '__init__{}'.format(extension)))
-                   for extension in PYTHON_EXTENSIONS):
+    Do we already discovered implementation?
+    What kind of module we discovered?
+    """
+    def __init__(self) -> None:
+        self.has_py = False  # type: bool
+        self.type = None  # type: Optional[ModuleType]
+        # Paths can contain only one ".py" path, but multiple stubs
+        self.paths = []  # type: List[str]
+
+    def maybe_add_path(self, path: str, type: ModuleType) -> None:
+        """
+        Add path to import context.
+        Modifies self.paths in case if arguments satisfy import context state
+        """
+        assert path.endswith((os.path.sep,) + tuple(PYTHON_EXTENSIONS))
+
+        if self.type is not None and self.type != type:
+            return None
+
+        # We can add at most one module implementation to paths
+        # But module can have multiple stubs
+        py_path = path.endswith('.py')
+        if self.has_py and py_path:
+            return None
+
+        if type == ModuleType.namespace:
+            ok = self._verify_namespace(path)
+        else:
+            ok = self._verify_module(path)
+
+        if not ok:
+            return None
+
+        if py_path:
+            self.has_py = True
+
+        self.type = type
+        self.paths.append(path)
+
+    def _verify_module(self, path: str) -> bool:
```

Comment:
We don't need this verification at all, `_collect_paths` won't ever try to add a path that it hasn't already verified exists.

Reply:
> We don't need this verification at all, _collect_paths won't ever try to add a path that it hasn't already verified exists.

That is incorrect we trying to add dynamically computed path and we don't know which of them exists at this moment yet. We are just trying all combinations in the specific order and hope that one of them fits.